### PR TITLE
docs: add documentation for Keras and PyTorch profilers [MLG-1094]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

Add/update user-facing documentation for native profilers (TF Keras and PyTorch). Features have existed for a long time, but documentation has been missing.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: docs/model-dev-guide/api-guides/apis-howto/api-keras-ug.rst [hunk 1]
```diff
@@ -94,3 +94,41 @@ implement the :class:`determined.keras.callbacks.Callback` interface (an extensi
...
+ Profiling
+***********
+
+Determined supports integration with the native TF Keras profiler. Results will automatically be
+uploaded to Tensorboard, which is accessible from trial page on the Determined UI.
+
+The Keras profiler is configured as a callback in the :class:`~determined.keras.TFKerasTrial` class.
+The :class:`determined.keras.callbacks.TensorBoard` callback is a thin wrapper around the native
+Keras Tensorboard callback, ``tf.keras.callbacks.TensorBoard``. It overrides the ``log_dir``
+argument to the Determined Tenosrboard path, while other arguments are passed directly into
```
Comment (tara-det-ai) [inline docs/model-dev-guide/api-guides/apis-howto/api-keras-ug.rst:108]: TensorBoard

File: docs/model-dev-guide/api-guides/apis-howto/api-keras-ug.rst [hunk 2]
```diff
@@ -94,3 +94,41 @@ implement the :class:`determined.keras.callbacks.Callback` interface (an extensi
 ``tf.keras.callbacks.Callbacks`` interface) and supply them to the
 :class:`~determined.keras.TFKerasTrial` by implementing
 :meth:`~determined.keras.TFKerasTrial.keras_callbacks`.
+
+***********
+ Profiling
+***********
+
+Determined supports integration with the native TF Keras profiler. Results will automatically be
+uploaded to Tensorboard, which is accessible from trial page on the Determined UI.
```
Comment (tara-det-ai) [inline docs/model-dev-guide/api-guides/apis-howto/api-keras-ug.rst:103]: WebUI?

File: docs/model-dev-guide/api-guides/apis-howto/api-keras-ug.rst [hunk 3]
```diff
@@ -94,3 +94,41 @@ implement the :class:`determined.keras.callbacks.Callback` interface (an extensi
...
+              histogram_freq=1,
+          )
+      ]
+
+.. note::
+
+   Though specifying batches to profile with ``profile_batch`` is optional, profiling every batch
+   may cause a large amount of data to be uploaded to Tensorboard. This may result in long rendering
+   times for Tensorboard and memory issues. For long-running experiments, it is recommended to
+   configure profiling only on desired batches.
```
Comment (tara-det-ai) [inline docs/model-dev-guide/api-guides/apis-howto/api-keras-ug.rst:134]: ***********
 Profiling
***********

Determined supports integration with the native TF Keras profiler. Results are automatically uploaded to TensorBoard, accessible from the trial page in the WebUI.

Configuring the Keras Profiler
==============================

The Keras profiler is configured as a callback in the :class:`~determined.keras.TFKerasTrial` class. The :class:`determined.keras.callbacks.TensorBoard` callback is a thin wrapper around the native Keras TensorBoard callback, :class:`tf.keras.callbacks.TensorBoard`. It overrides the ``log_dir`` argument to set the Determined TensorBoard path, while other arguments are passed directly into :class:`tf.keras.callbacks.TensorBoard`. For a list of accepted arguments, consult the `official Keras API documentation
<https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/TensorBoard>`_.

Example Code Snippet
====================

This code snippet shows you how to configure profiling for 
batches 5 and 10, computing weight histograms every 1 epochs.

.. code-block:: python

   from determined import keras
   from typing import List
   import tensorflow as tf

   def keras_callbacks(self) -> List[tf.keras.callbacks.Callback]:
       return [
           keras.callbacks.TensorBoard(
               update_freq="batch",
               profile_batch='5, 10',
               histogram_freq=1,
           )
       ]

.. note::

   While specifying batches to profile with ``profile_batch`` is optional, profiling every batch can generate a large amount of data, causing long rendering times and potential memory issues in TensorBoard. For long-running experiments, it's advised to profile only specific batches.

File: docs/model-dev-guide/api-guides/apis-howto/api-pytorch-ug.rst [hunk 4]
```diff
@@ -549,6 +549,13 @@ See the `PyTorch tensorboard profiler tutorial
 <https://pytorch.org/tutorials/intermediate/tensorboard_profiler_tutorial.html#use-profiler-to-record-execution-events>`_
 for a complete list of accepted configurations parameters.
 
+.. note::
+
+   Though configuring a profiling schedule ``torch.profiler.schedule`` is optional, profiling every
+   batch may cause a large amount of data to be uploaded to Tensorboard. This may result in long
+   rendering times for Tensorboard and memory issues. For long-running experiments, it is
+   recommended to configure a profiling schedule.
+
```
Comment (tara-det-ai) [inline docs/model-dev-guide/api-guides/apis-howto/api-pytorch-ug.rst:558]: we could reuse the same note from above

**Review by tara-det-ai — APPROVED** [review]
suggested edits

**Comment by ioga** [comment]
@azhou-determined do you plan to also add the docs on how to use the native profilers in Core API?

File: docs/model-dev-guide/api-guides/apis-howto/api-core-ug.rst [hunk 5]
```diff
@@ -513,6 +513,56 @@ In the Determined WebUI, go to the **Cluster** pane.
 You should be able to see multiple slots active corresponding to the value you set for
 ``slots_per_trial`` you set in ``distributed.yaml``, as well as logs appearing from multiple ranks.
 
+***********
+ Profiling
```
Comment (azhou-determined) [inline docs/model-dev-guide/api-guides/apis-howto/api-core-ug.rst:517]: test comment

Comment (azhou-determined) [reply docs/model-dev-guide/api-guides/apis-howto/api-core-ug.rst:517]: @azhou-determined  mention
